07_errors/nicholascross/main.go

```diff
@@ -0,0 +1,169 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+type Puppy struct {
+	ID     int64
+	Breed  string
+	Colour string
+	Value  float64
+}
+
+type Storer interface {
+	CreatePuppy(p Puppy) (bool, error)
+	RetrievePuppy(int64) (*Puppy, error)
+	UpdatePuppy(int64, Puppy) (bool, error)
+	DeletePuppy(int64) (bool, error)
+}
+
+type MapStore map[int64]Puppy
+
+const (
+	missingPup = iota
+	invalidPupValue
+	pupAlreadyExists
+)
+
+type PupError struct {
+	Message string
+	Code    int
```

Comment:
Rather than one error type with a code to differentiate the errors, Go tends towards defining different error types for different errors named `FooError` for parameterised errors, or package-level variables that are error values created with `errors.New()` named `ErrFoo` for static errors.

See https://github.com/golang/go/blob/master/src/encoding/json/encode.go#L230 for an example of the former and https://github.com/golang/go/blob/master/src/encoding/gob/decode.go#L19 for an example of the latter.

Or just create them on the fly as needed: https://github.com/golang/go/blob/master/src/database/sql/sql.go#L542

https://github.com/golang/go/wiki/Errors also has a bit more to say about this too.

I think the main thing is that your errors don't look like errors (from a Go perspective). `return nil, missingPuppy(id)` does not say err(or) anywhere which is what I would usually expect to see in Go code.

Reply:
Ok, so it looks like I didn't read the lab description where it says "Add a custom error type `Error` with fields `Message` and `Code`" - so ignore my comment about using `Code` (or not, as you please). I think my last paragraph in this comment still stands though - the errors don't look like errors, so if you can put "[Ee]rr" in the names somewhere I think it would read better.